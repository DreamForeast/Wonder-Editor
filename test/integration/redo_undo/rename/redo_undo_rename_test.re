open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe("redo_undo: rename", () => {
    let sandbox = getSandboxDefaultVal();
    beforeEach(() => {
      sandbox := createSandbox();
      StateHistoryToolEditor.clearAllState();
    });
    let _simulateTwiceChangeName = () => {
      let inspectorComponent =
        BuildComponentTool.buildInspectorComponent(
          TestTool.buildAppStateSceneGraphFromEngine(),
          InspectorTool.buildFakeAllShowComponentConfig(),
        );
      let name1 = "gameObject1";
      BaseEventTool.triggerComponentEvent(
        inspectorComponent,
        GameObjectRenameTool.triggerRenameChangeEvent(name1),
      );
      BaseEventTool.triggerComponentEvent(
        inspectorComponent,
        GameObjectRenameTool.triggerRenameBlurEvent(name1),
      );
      let name2 = "gameObject2";
      BaseEventTool.triggerComponentEvent(
        inspectorComponent,
        GameObjectRenameTool.triggerRenameChangeEvent(name2),
      );
      BaseEventTool.triggerComponentEvent(
        inspectorComponent,
        GameObjectRenameTool.triggerRenameBlurEvent(name2),
      );
    };

    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));
    describe("test simulate set currentSceneTreeNode", () => {
      beforeEach(() => {
        TestTool.closeContractCheck();
        MainEditorSceneTool.initStateAndGl(~sandbox, ());
        MainEditorSceneTool.createDefaultScene(sandbox, () => ());

        SceneTreeNodeDomTool.OperateTwoLayer.getFirstCubeDomIndex()
        |> SceneTreeTool.clearCurrentGameObjectAndSetTreeSpecificGameObject;
      });
      afterEach(() => TestTool.openContractCheck());

      describe("test undo operate", () => {
        test("test not undo", () => {
          _simulateTwiceChangeName();

          BuildComponentTool.buildInspectorComponent(
            TestTool.buildAppStateSceneGraphFromEngine(),
            InspectorTool.buildFakeAllShowComponentConfig(),
          )
          |> ReactTestTool.createSnapshotAndMatch;
        });
        describe("test undo one step", () => {
          test("step which from second to first", () => {
            _simulateTwiceChangeName();

            StateHistoryToolEditor.undo();

            BuildComponentTool.buildInspectorComponent(
              TestTool.buildAppStateSceneGraphFromEngine(),
              InspectorTool.buildFakeAllShowComponentConfig(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          });
          test("step which from second to zero", () => {
            _simulateTwiceChangeName();

            StateHistoryToolEditor.undo();
            StateHistoryToolEditor.undo();

            BuildComponentTool.buildInspectorComponent(
              TestTool.buildAppStateSceneGraphFromEngine(),
              InspectorTool.buildFakeAllShowComponentConfig(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          });
        });
      });
      describe("test redo operate", () => {
        describe("test redo one step", () => {
          test("if not exec undo, redo one step, not change", () => {
            _simulateTwiceChangeName();

            StateHistoryToolEditor.redo();

            BuildComponentTool.buildInspectorComponent(
              TestTool.buildAppStateSceneGraphFromEngine(),
              InspectorTool.buildFakeAllShowComponentConfig(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          });
          test(
            "undo step which from second to zero, redo step which from zero to first",
            () => {
            _simulateTwiceChangeName();

            StateHistoryToolEditor.undo();
            StateHistoryToolEditor.undo();
            StateHistoryToolEditor.redo();

            BuildComponentTool.buildInspectorComponent(
              TestTool.buildAppStateSceneGraphFromEngine(),
              InspectorTool.buildFakeAllShowComponentConfig(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          });
        });
        describe("test redo two step", () =>
          test(
            "undo step which from second to zero, redo step which from zero to second",
            () => {
            _simulateTwiceChangeName();

            StateHistoryToolEditor.undo();
            StateHistoryToolEditor.undo();
            StateHistoryToolEditor.redo();
            StateHistoryToolEditor.redo();

            BuildComponentTool.buildInspectorComponent(
              TestTool.buildAppStateSceneGraphFromEngine(),
              InspectorTool.buildFakeAllShowComponentConfig(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          })
        );
        describe("test redo three step", () =>
          test(
            "test if current step is last step, execute redo, not change", () => {
            _simulateTwiceChangeName();

            StateHistoryToolEditor.undo();
            StateHistoryToolEditor.undo();
            StateHistoryToolEditor.redo();
            StateHistoryToolEditor.redo();
            StateHistoryToolEditor.redo();

            BuildComponentTool.buildInspectorComponent(
              TestTool.buildAppStateSceneGraphFromEngine(),
              InspectorTool.buildFakeAllShowComponentConfig(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          })
        );
      });
    });
  });