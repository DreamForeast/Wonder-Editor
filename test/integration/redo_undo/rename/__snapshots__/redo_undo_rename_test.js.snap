// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redo_undo: rename test simulate set currentSceneTreeNode test redo operate test redo one step if not exec undo, redo one step, not change 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-sceneTree"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          Name
        </div>
      </div>
      <div>
        <article
          className="wonder-string-input"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="gameObject2"
          />
        </article>
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="wonder-inspector-transform"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="wonder-inspector-material"
      >
        <article
          className="wonder-material-color"
        >
          <div
            className=""
          >
            <span
              className=""
            >
              color : 
            </span>
            <span
              className=""
            >
              #ffffff
            </span>
            <button
              className=""
              onClick={[Function]}
            >
              pick color
            </button>
          </div>
        </article>
        <article
          className="wonder-material-texture"
        >
          <div
            className="texture_ground"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "opacity": "1",
              }
            }
          />
          <span
            className=""
          >
            texture:
          </span>
          <img
            src="./public/img/null.jpg"
          />
          <button
            className="texture_remove"
            onClick={[Function]}
          >
            remove
          </button>
        </article>
      </article>
    </article>
    <article
      className="wonder-addable-component"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`redo_undo: rename test simulate set currentSceneTreeNode test redo operate test redo one step undo step which from second to zero, redo step which from zero to first 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-sceneTree"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          Name
        </div>
      </div>
      <div>
        <article
          className="wonder-string-input"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="gameObject1"
          />
        </article>
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="wonder-inspector-transform"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="wonder-inspector-material"
      >
        <article
          className="wonder-material-color"
        >
          <div
            className=""
          >
            <span
              className=""
            >
              color : 
            </span>
            <span
              className=""
            >
              #ffffff
            </span>
            <button
              className=""
              onClick={[Function]}
            >
              pick color
            </button>
          </div>
        </article>
        <article
          className="wonder-material-texture"
        >
          <div
            className="texture_ground"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "opacity": "1",
              }
            }
          />
          <span
            className=""
          >
            texture:
          </span>
          <img
            src="./public/img/null.jpg"
          />
          <button
            className="texture_remove"
            onClick={[Function]}
          >
            remove
          </button>
        </article>
      </article>
    </article>
    <article
      className="wonder-addable-component"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`redo_undo: rename test simulate set currentSceneTreeNode test redo operate test redo three step test if current step is last step, execute redo, not change 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-sceneTree"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          Name
        </div>
      </div>
      <div>
        <article
          className="wonder-string-input"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="gameObject2"
          />
        </article>
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="wonder-inspector-transform"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="wonder-inspector-material"
      >
        <article
          className="wonder-material-color"
        >
          <div
            className=""
          >
            <span
              className=""
            >
              color : 
            </span>
            <span
              className=""
            >
              #ffffff
            </span>
            <button
              className=""
              onClick={[Function]}
            >
              pick color
            </button>
          </div>
        </article>
        <article
          className="wonder-material-texture"
        >
          <div
            className="texture_ground"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "opacity": "1",
              }
            }
          />
          <span
            className=""
          >
            texture:
          </span>
          <img
            src="./public/img/null.jpg"
          />
          <button
            className="texture_remove"
            onClick={[Function]}
          >
            remove
          </button>
        </article>
      </article>
    </article>
    <article
      className="wonder-addable-component"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`redo_undo: rename test simulate set currentSceneTreeNode test redo operate test redo two step undo step which from second to zero, redo step which from zero to second 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-sceneTree"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          Name
        </div>
      </div>
      <div>
        <article
          className="wonder-string-input"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="gameObject2"
          />
        </article>
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="wonder-inspector-transform"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="wonder-inspector-material"
      >
        <article
          className="wonder-material-color"
        >
          <div
            className=""
          >
            <span
              className=""
            >
              color : 
            </span>
            <span
              className=""
            >
              #ffffff
            </span>
            <button
              className=""
              onClick={[Function]}
            >
              pick color
            </button>
          </div>
        </article>
        <article
          className="wonder-material-texture"
        >
          <div
            className="texture_ground"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "opacity": "1",
              }
            }
          />
          <span
            className=""
          >
            texture:
          </span>
          <img
            src="./public/img/null.jpg"
          />
          <button
            className="texture_remove"
            onClick={[Function]}
          >
            remove
          </button>
        </article>
      </article>
    </article>
    <article
      className="wonder-addable-component"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`redo_undo: rename test simulate set currentSceneTreeNode test undo operate test not undo 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-sceneTree"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          Name
        </div>
      </div>
      <div>
        <article
          className="wonder-string-input"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="gameObject2"
          />
        </article>
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="wonder-inspector-transform"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="wonder-inspector-material"
      >
        <article
          className="wonder-material-color"
        >
          <div
            className=""
          >
            <span
              className=""
            >
              color : 
            </span>
            <span
              className=""
            >
              #ffffff
            </span>
            <button
              className=""
              onClick={[Function]}
            >
              pick color
            </button>
          </div>
        </article>
        <article
          className="wonder-material-texture"
        >
          <div
            className="texture_ground"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "opacity": "1",
              }
            }
          />
          <span
            className=""
          >
            texture:
          </span>
          <img
            src="./public/img/null.jpg"
          />
          <button
            className="texture_remove"
            onClick={[Function]}
          >
            remove
          </button>
        </article>
      </article>
    </article>
    <article
      className="wonder-addable-component"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`redo_undo: rename test simulate set currentSceneTreeNode test undo operate test undo one step step which from second to first 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-sceneTree"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          Name
        </div>
      </div>
      <div>
        <article
          className="wonder-string-input"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="gameObject1"
          />
        </article>
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="wonder-inspector-transform"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="wonder-inspector-material"
      >
        <article
          className="wonder-material-color"
        >
          <div
            className=""
          >
            <span
              className=""
            >
              color : 
            </span>
            <span
              className=""
            >
              #ffffff
            </span>
            <button
              className=""
              onClick={[Function]}
            >
              pick color
            </button>
          </div>
        </article>
        <article
          className="wonder-material-texture"
        >
          <div
            className="texture_ground"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "opacity": "1",
              }
            }
          />
          <span
            className=""
          >
            texture:
          </span>
          <img
            src="./public/img/null.jpg"
          />
          <button
            className="texture_remove"
            onClick={[Function]}
          >
            remove
          </button>
        </article>
      </article>
    </article>
    <article
      className="wonder-addable-component"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`redo_undo: rename test simulate set currentSceneTreeNode test undo operate test undo one step step which from second to zero 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-sceneTree"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          Name
        </div>
      </div>
      <div>
        <article
          className="wonder-string-input"
        >
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="cube"
          />
        </article>
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="wonder-inspector-transform"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="wonder-inspector-material"
      >
        <article
          className="wonder-material-color"
        >
          <div
            className=""
          >
            <span
              className=""
            >
              color : 
            </span>
            <span
              className=""
            >
              #ffffff
            </span>
            <button
              className=""
              onClick={[Function]}
            >
              pick color
            </button>
          </div>
        </article>
        <article
          className="wonder-material-texture"
        >
          <div
            className="texture_ground"
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            style={
              Object {
                "opacity": "1",
              }
            }
          />
          <span
            className=""
          >
            texture:
          </span>
          <img
            src="./public/img/null.jpg"
          />
          <button
            className="texture_remove"
            onClick={[Function]}
          >
            remove
          </button>
        </article>
      </article>
    </article>
    <article
      className="wonder-addable-component"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;
